cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/bonded.h>
+#include <grpc++/impl/codegen/status.h>
+
+#include <boost/optional.hpp>
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <tuple>
+
+namespace bond { namespace ext { namespace gRPC {
+
+/// @brief A callback type that can be manually waited upon.
+///
+/// The type can be used to synchronously get the result of invoking an
+/// async proxy method.
+///
+/// The wait() member function can be used to wait until the callback has
+/// been called. Then, the status() and response() member functions can be
+/// called to inspect the results.
+template <typename TResponse>
+class wait_callback
+{
+public:
+    wait_callback() : _impl(std::make_shared<impl>()) { }
+
+    /// @brief Records the response and status.
+    ///
+    /// May be invoked multiple times and from different threads, but only the first
+    /// invocation will be recorded.
```

Comment:
Is this design approach intentional and required? I agree with `std::promise::set_value` which throws exception on second call and does not hide issues from user.